### PR TITLE
Complete typespecs and guards for Module module

### PR DESCRIPTION
**Pull request by eksperimental**
There are plenty of changes, that's why the [WIP], but it all guards and typespecs are complete for this module.

There is a common argument which is the `{<function/macro atom>, <arity>}` tuple,
and it make me think whether we need to create a new `type` for this.

### CODE REVIEW

File: lib/elixir/lib/module.ex [hunk 1]
```diff
@@ -725,7 +733,7 @@ defmodule Module do
       end
 
   """
-  def defines?(module, tuple) when is_tuple(tuple) do
+  def defines?(module, tuple) when is_atom(module) and is_tuple(tuple) and tuple_size(tuple) == 2 do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Let's do `defines?(module, {_, _} = tuple) when is_atom(module)`.

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -749,7 +757,9 @@ defmodule Module do
       end
 
   """
-  def defines?(module, tuple, kind) do
+  @spec defines?(module, {function_name :: atom, arity}, :def | :defp | :defmacro | :defmacrop) :: boolean
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Let's do `defines?(module, {_, _} = tuple, ...) when is_atom(module)`.

**Comment by josevalim** [comment]
@eksperimental yup, a new type would be super helpful here.

File: lib/elixir/lib/module.ex (hunk 1 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: great!
besides being neater, does it perform faster?

Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

File: lib/elixir/lib/module.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed and improved

File: lib/elixir/lib/module.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/module.ex]: perf diff is negligent.

Comment (eksperimental) [inline lib/elixir/lib/module.ex]: do you mean _negligible_?

Comment (josevalim) [inline lib/elixir/lib/module.ex]: hahahahaha, yes

**Comment by eksperimental** [comment]
@josevalim so what should we name it?

`@type function_macro_name_arity :: {atom, arity}` ?

**Comment by josevalim** [comment]
@eksperimental `@type function_arity :: {function, arity}`.

**Comment by eksperimental** [comment]
you mean `atom` not `function`, right?

**Comment by josevalim** [comment]
I meant `function`, it is an actual built-in type that is a synonym to atom.

**Comment by eksperimental** [comment]
@josevalim that is not documented in Elixir, and in Erlang
http://erlang.org/doc/reference_manual/typespec.html
`function()` is defined as `fun()`.

**Comment by fishcakez** [comment]
@eksperimental IIRC it didn't used to be documented in Erlang and when it wasn't documented it meant `atom()`. Then some people pointed this out, and it got changed to `fun()` and documented.

**Comment by josevalim** [comment]
So let's go with `atom()` to avoid confusion.

**Comment by eksperimental** [comment]
thank you @fishcakez, interesting and ambiguous.
I am trying to see when this change happened in OTP 18 or later, to see if we can include it in the typespecs page

**Comment by eksperimental** [comment]
So it is documented at since OTP-17.4
https://github.com/erlang/otp/commit/b7f72c888fe20fb7e4ade1c68647d7be92d0f427
So I guess we can document it in Elixir as well.
I will submit a PR.
thank you @fishcakez for the clarification

**Comment by josevalim** [comment]
Please let us know when this is ready for another pass.

**Comment by eksperimental** [comment]
sorry @josevalim,
this is ready for review since 3 days ago

File: lib/elixir/lib/module.ex [hunk 3]
```diff
@@ -1015,7 +1040,8 @@ defmodule Module do
   @doc false
   # Used internally to compile types.
   # This function is private and must be used only internally.
-  def store_typespec(module, key, value) when is_atom(key) do
+  @spec store_typespec(module, key :: atom, value :: term) :: Keyword.t
```
Comment (fishcakez) [inline lib/elixir/lib/module.ex]: This function returns `true`, not `Keyword.t`.

Personally I think adding type checking guards to internal functions like this one is too defensive. Normally type check on boundary and use guards for flow control internally.

Is there reason spec'ed this undocumented function and not the other two (get/put attribute)?

File: lib/elixir/lib/module.ex [hunk 4]
```diff
@@ -463,7 +467,7 @@ defmodule Module do
   """
   def create(module, quoted, opts)
 
-  def create(module, quoted, %Macro.Env{} = env) do
+  def create(module, quoted, %Macro.Env{} = env) when is_atom(module) do
```
Comment (fishcakez) [inline lib/elixir/lib/module.ex]: `create/3` is missing spec

File: lib/elixir/lib/module.ex [hunk 5]
```diff
@@ -550,18 +554,19 @@ defmodule Module do
 
   """
   @spec safe_concat(binary | atom, binary | atom) :: atom | no_return
-  def safe_concat(left, right) do
```
Comment (fishcakez) [inline lib/elixir/lib/module.ex]: This should be `:: atom`, no `no_return`, could you fix this as you are fixing the other specs?

File: lib/elixir/lib/module.ex [hunk 6]
```diff
@@ -788,7 +806,8 @@ defmodule Module do
       end
 
   """
-  def definitions_in(module, kind) do
+  @spec definitions_in(module, atom) :: [function_arity]
```
Comment (fishcakez) [inline lib/elixir/lib/module.ex]: Is atom too general? Perhaps the different kinds could have their own types and reused throughout the specs?

File: lib/elixir/lib/module.ex [hunk 7]
```diff
@@ -801,14 +820,16 @@ defmodule Module do
   developer to customize it. See `Kernel.defoverridable/1` for
   more information and documentation.
   """
-  def make_overridable(module, tuples) do
+  @spec make_overridable(module, [function_arity]) :: :ok | no_return
```
Comment (fishcakez) [inline lib/elixir/lib/module.ex]: Should be `:: :ok`, no `no_return`

File: lib/elixir/lib/module.ex [hunk 8]
```diff
@@ -970,6 +994,7 @@ defmodule Module do
       ["Very", "Long", "Module", "Name", "And", "Even", "Longer"]
 
   """
+  @spec split(module) :: [String.t]
```
Comment (fishcakez) [inline lib/elixir/lib/module.ex]: Will return non-empty list of strings: `[String.t, ...]`

**Comment by fishcakez** [comment]
So the guards will cause an immediate function clause error on bad
arguments. This should help users fix bugs in their code without needing to
read the source of elixir. It is especially to improve experience for new
users who won't be familiar with debugging elixir or know where to look.

File: lib/elixir/lib/module.ex (hunk 3 above)
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Since this is `@doc false`, I agree. No need or specs nor guards.

Comment (eksperimental) [inline lib/elixir/lib/module.ex]: spec removed

File: lib/elixir/lib/module.ex (hunk 8 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

File: lib/elixir/lib/module.ex (hunk 7 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

File: lib/elixir/lib/module.ex (hunk 6 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: yes. it's a good idea. done!

File: lib/elixir/lib/module.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

File: lib/elixir/lib/module.ex (hunk 4 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: done. please review it

**Comment by eksperimental** [comment]
suggestions added, please review them.

File: lib/elixir/lib/module.ex [hunk 9]
```diff
@@ -725,7 +738,10 @@ defmodule Module do
       end
 
   """
-  def defines?(module, tuple) when is_tuple(tuple) do
+  @spec defines?(module, function_arity) :: boolean
+  def defines?(module, {function_macro_name, arity} = tuple)
+      when is_atom(module) and is_atom(function_macro_name)
+      and is_integer(arity) and arity >= 0 and arity <= 255 do
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: Please put the `and` at the end of the previous line, and indent `is_integer` to be aligned with the `is_atom` on the line above:

```elixir
def ...
    when is_atom(module) and is_atom(function_macro_name) and
         is_integer(arity) and arity >= 0 ... do
```

File: lib/elixir/lib/module.ex [hunk 10]
```diff
@@ -749,11 +765,15 @@ defmodule Module do
       end
 
   """
-  def defines?(module, tuple, kind) do
+  @spec defines?(module, function_arity, def_kind) :: boolean
+  def defines?(module, {function_macro_name, arity} = tuple, def_kind)
+      when is_atom(module) and is_atom(function_macro_name)
+      and is_integer(arity) and arity >= 0 and arity <= 255
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: Same as https://github.com/elixir-lang/elixir/pull/5869/files#r106882072.

File: lib/elixir/lib/module.ex (hunk 7 above)
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: There's still the `no_return` here 😕

**Comment by whatyouhide** [comment]
I honestly feel a bit like @fishcakez, this PR is quite "heavy" on the verboseness side. I understand the point in doing this but I can't shake off the bad feeling :bowtie:

**Comment by michalmuskala** [comment]
To be honest, I'm not very keen on adding guards everywhere. They should be considered more closely, since they don't always improve errors.

For example with `Enum.map` where there was no function guard, when you passed wrong arity, you'd get `BadArityError` and now you get a generic `FunctionClauseError` - for me, the error is much worse now, with the guards, the old one pointed at the issue much more directly.

**Comment by josevalim** [comment]
@michalmuskala we should revisit such cases then but I would say that's rather the exception than the rule. We should probably consider reverting the function guards there. Other cases where we typically raise good messages is when using maps.

Other than that, it is most likely that the error message will be bad. For example, an unchecked tuple can potentially raise badarg when calling element three or for function calls inside. That's arguably a worst experience.

For what is worth, most of the API in OTP is correctly guarded. In Elixir we never had the habit.

So overall the rationale is to provide better documentation and error messages. I agree the system is verbose and I don't like it either but if it improves the language experience with more documentation and error messages, then I think it is worth it.

**Comment by eksperimental** [comment]
@whatyouhide corrections made

File: lib/elixir/lib/module.ex [hunk 11]
```diff
@@ -801,14 +823,16 @@ defmodule Module do
...
   more information and documentation.
   """
-  def make_overridable(module, tuples) do
+  @spec make_overridable(module, [function_arity]) :: :ok
+  def make_overridable(module, tuples) when is_atom(module) and is_list(tuples) do
     assert_not_compiled!(:make_overridable, module)
 
-    :lists.foreach(fn {name, arity} = tuple ->
+    :lists.foreach(fn {function_name, arity} = tuple
+        when is_atom(function_name) and is_integer(arity) and arity >= 0 and arity <= 255 ->
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: I don't think this is a helpful error message. Blowing up with "no function clause matching" is gonna be pretty obscure since this comes from the guts of this function. I think here it could be beneficial to explicitly handle bad tuples:

```elixir
:lists.foreach(fn
  {function_name, arity} = tuple when is_atom(function_name) and is_integer(arity) and arity >= 0 and arity <= 255 ->
    ...
  other ->
    raise ArgumentError, "each element has to be a {function_name :: atom, arity :: 1..255} tuple, got: #{inspect(other)}"
end, ...)
```

Comment (eksperimental) [inline lib/elixir/lib/module.ex]: I totally agree!
done

File: lib/elixir/lib/module.ex [hunk 12]
```diff
@@ -782,16 +803,17 @@ defmodule Module do
   ## Examples
 
       defmodule Example do
-        def version, do: 1
+        defmacro version, do: 1
         Module.definitions_in __MODULE__, :def  #=> [{:version, 0}]
-        Module.definitions_in __MODULE__, :defp #=> []
+        Module.definitions_in __MODULE__, :defmacro #=> []
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: Why this change? If you change `version/0` to be a macro, then `Module.definitions_in(__MODULE__, :def)` will return `[]`, not `[{:version, 0}]`, and `:defmacro` will return `[{:version, 0}]` instead 😕

File: lib/elixir/lib/module.ex [hunk 13]
```diff
@@ -782,16 +803,17 @@ defmodule Module do
...
       end
 
   """
-  def definitions_in(module, kind) do
+  @spec definitions_in(module, def_kind) :: [function_arity]
+  def definitions_in(module, def_kind) when is_atom(module) and is_atom(def_kind) do
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: I think we should have the usual `when def_kind in [:def, defp, ...]` here.

File: lib/elixir/lib/module.ex [hunk 14]
```diff
@@ -801,42 +823,51 @@ defmodule Module do
...
-            "cannot make function #{name}/#{arity} overridable because it was not defined"
-        clause ->
-          neighbours =
-            if :elixir_compiler.get_opt(:internal) do
-              []
-            else
-              Module.LocalsTracker.yank(module, tuple)
+    :lists.foreach(fn
+      {function_name, arity} = tuple
+          when is_atom(function_name) and is_integer(arity) and arity >= 0 and arity <= 255 ->
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: Please indent this on the same line as the lone above. It's not that long, but this indentation looks a bit off :bowtie:

File: lib/elixir/lib/module.ex [hunk 15]
```diff
@@ -801,42 +823,51 @@ defmodule Module do
...
+              "cannot make function #{function_name}/#{arity} overridable because it was not defined"
+          clause ->
+            neighbours =
+              if :elixir_compiler.get_opt(:internal) do
+                []
+              else
+                Module.LocalsTracker.yank(module, tuple)
+              end
+
+            old   = :elixir_overridable.overridable(module)
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: Can we drop the `=` alignment since we're touching these lines?

File: lib/elixir/lib/module.ex [hunk 16]
```diff
@@ -879,8 +910,8 @@ defmodule Module do
       end
 
   """
-  @spec get_attribute(module, atom) :: term
-  def get_attribute(module, key) do
+  @spec get_attribute(module, key :: atom) :: (value :: term)
```
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: I don't think we need to name the types here (`key ::` and `value ::`), they're clear from the function arguments and the function docs.

File: lib/elixir/lib/module.ex (hunk 14 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: do you mean move it to the line above, or indent it to the same level as the line above?

File: lib/elixir/lib/module.ex (hunk 12 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: I was probably figuring out the spec and played with the example and accidentally saved it.

File: lib/elixir/lib/module.ex (hunk 13 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

File: lib/elixir/lib/module.ex [hunk 17]
```diff
@@ -351,6 +351,10 @@ defmodule Module do
   the documentation for the [`:compile` module](http://www.erlang.org/doc/man/compile.html).
   '''
 
+  @type function_arity :: {atom, arity}
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Let's make those `typep` as I don't want others to depend on it.

File: lib/elixir/lib/module.ex [hunk 18]
```diff
@@ -571,22 +579,27 @@ defmodule Module do
...
+  def add_doc(_module, _line, kind, _function_tuple, _signature, doc)
+      when kind in [:defp, :defmacrop, :typep] do
     if doc, do: {:error, :private_doc}, else: :ok
   end
 
-  def add_doc(module, line, kind, tuple, signature, doc) when
-      kind in [:def, :defmacro, :type, :opaque] and (is_binary(doc) or is_boolean(doc) or doc == nil) do
+  def add_doc(module, line, kind, function_tuple, signature, doc)
+      when kind in [:def, :defmacro, :type, :opaque]
+      and (is_binary(doc) or is_boolean(doc) or doc == nil) do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: And should be in the previous line and the parens should be aligned with the beginning of `kind`.

File: lib/elixir/lib/module.ex [hunk 19]
```diff
@@ -749,11 +765,15 @@ defmodule Module do
       end
 
   """
-  def defines?(module, tuple, kind) do
+  @spec defines?(module, function_arity, def_kind) :: boolean
+  def defines?(module, {function_macro_name, arity} = tuple, def_kind)
+      when is_atom(module) and is_atom(function_macro_name) and
+           is_integer(arity) and arity >= 0 and arity <= 255 and
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Doesn't `arity in 0..255` work here?

File: lib/elixir/lib/module.ex [hunk 20]
```diff
@@ -801,42 +824,50 @@ defmodule Module do
...
-          count = case :maps.find(tuple, old) do
-            {:ok, {count, _, _, _}} -> count + 1
-            :error -> 1
-          end
-          new = :maps.put(tuple, {count, clause, neighbours, false}, old)
-          :elixir_overridable.overridable(module, new)
-      end
+      other ->
+        raise ArgumentError,
+              "each element in tuple list has to be a {function_name :: atom, arity :: 1..255} tuple, got: #{inspect(other)}"
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Arity is from 0 to 255.

File: lib/elixir/lib/module.ex [hunk 21]
```diff
@@ -801,42 +824,50 @@ defmodule Module do
...
     end, tuples)
   end
 
   @doc """
   Returns `true` if `tuple` in `module` is marked as overridable.
   """
-  def overridable?(module, tuple) do
+  @spec overridable?(module, function_arity) :: boolean
+  def overridable?(module, {function_name, arity} = tuple)
+      when is_atom(function_name) and is_integer(arity) and arity >= 0 and arity <= 255 do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: `arity in 0..255`?

File: lib/elixir/lib/module.ex [hunk 22]
```diff
@@ -879,8 +910,8 @@ defmodule Module do
       end
 
   """
-  @spec get_attribute(module, atom) :: term
-  def get_attribute(module, key) do
+  @spec get_attribute(module, atom) :: (term)
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Please remove parens around term.

File: lib/elixir/lib/module.ex [hunk 23]
```diff
@@ -897,8 +928,8 @@ defmodule Module do
       end
 
   """
-  @spec delete_attribute(module, key :: atom) :: (value :: term)
-  def delete_attribute(module, key) when is_atom(key) do
+  @spec delete_attribute(module, atom) :: (term)
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Please remove parens around term.

File: lib/elixir/lib/module.ex [hunk 24]
```diff
@@ -944,18 +975,20 @@ defmodule Module do
       end
 
   """
-  def register_attribute(module, new, opts) when is_atom(new) do
+  @spec register_attribute(module, attribute :: atom, opts :: [{:accumulate, boolean}, {:persist, boolean}])
+      :: :ok | no_return
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Remove `no_return` from here per the previous discussions.

File: lib/elixir/test/elixir/module_test.exs [hunk 25]
```diff
@@ -310,4 +316,33 @@ defmodule ModuleTest do
       assert Module.definitions_in(__MODULE__, :defp) == []
     end
   end
+
+  describe "make_overridable/2 arguments" do
```
Comment (josevalim) [inline lib/elixir/test/elixir/module_test.exs]: Call it `"make_overridable/2"`.

Comment (josevalim) [inline lib/elixir/test/elixir/module_test.exs]: Check lib/elixir/test/elixir/kernel/overridable_test.exs. We don't need to add the success test because there are many there. You can add a test for the error in make_overridable though.

File: lib/elixir/lib/module.ex (hunk 19 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: Nope. They wouldn't work at this point. Ranges work different than lists with `in`.

File: lib/elixir/lib/module.ex (hunk 20 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: thank you. fixed

File: lib/elixir/lib/module.ex (hunk 21 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: It won't work.

File: lib/elixir/lib/module.ex (hunk 22 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: thank you

File: lib/elixir/lib/module.ex (hunk 23 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: done

File: lib/elixir/lib/module.ex (hunk 24 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: done

File: lib/elixir/lib/module.ex (hunk 18 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

File: lib/elixir/lib/module.ex (hunk 17 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: Changed `function_arity`, `def_kind`, `type_kind` to private.

File: lib/elixir/test/elixir/module_test.exs (hunk 25 above)
Comment (eksperimental) [inline lib/elixir/test/elixir/module_test.exs]: Thank you. have a look please.

**Comment by eksperimental** [comment]
@josevalim All your suggestions have been implemented

**Comment by josevalim** [comment]
Sorry but I can't see why the range version wouldn't work. arity in 0..255
expands to exactly the same code you wrote.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by eksperimental** [comment]
Well, I originally did it that way and I had to change it because it wouldn't work.

(second line is the offending line no. 743)
```elixir
  def defines?(module, {function_macro_name, arity} = tuple)
      when is_atom(module) and is_atom(function_macro_name) and arity in 0..255 do
    assert_not_compiled!(:defines?, module)
    table = defs_table_for(module)
    :ets.lookup(table, {:def, tuple}) != []
  end
```

`$ make clean test` will error:

```
...
==> bootstrap (compile)
Compiled lib/elixir/lib/kernel.ex
Compiled lib/elixir/lib/macro/env.ex
Compiled lib/elixir/lib/keyword.ex
error: undef
stacktrace: [{'Elixir.Macro',to_string,[{'..',[{line,743}],[0,255]}],[]},
             {'Elixir.Kernel',in,2,[{file,"expanding macro"}]},
             {'Elixir.Module','defines?',2,
                              [{file,"/home/eksperimental/elixir/lib/elixir/lib/module.ex"},
                               {line,743}]},
             {'Elixir.Kernel','and',2,[{file,"expanding macro"}]},
             {'Elixir.Module','defines?',2,
                              [{file,"/home/eksperimental/elixir/lib/elixir/lib/module.ex"},
                               {line,743}]},
             {elixir,eval_forms,4,[{file,"src/elixir.erl"},{line,216}]},
             {elixir_compiler,eval_forms,3,
                              [{file,"src/elixir_compiler.erl"},{line,61}]},
             {elixir_compiler,quoted,3,
                              [{file,"src/elixir_compiler.erl"},{line,27}]}]
Makefile:78: recipe for target 'lib/elixir/ebin/Elixir.Kernel.beam' failed
```

**Comment by josevalim** [comment]
Ah, that's because of bootstrapping. Ok. Let's go as is.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

File: lib/elixir/lib/module.ex [hunk 26]
```diff
@@ -725,7 +738,10 @@ defmodule Module do
       end
 
   """
-  def defines?(module, tuple) when is_tuple(tuple) do
+  @spec defines?(module, function_arity) :: boolean
+  def defines?(module, {function_macro_name, arity} = tuple)
```
Comment (lexmag) [inline lib/elixir/lib/module.ex]: `function_macro_name` seems quite confusing to me, can we just say `function_name`?

File: lib/elixir/lib/module.ex [hunk 27]
```diff
@@ -749,11 +765,15 @@ defmodule Module do
...
+  @spec defines?(module, function_arity, def_kind) :: boolean
+  def defines?(module, {function_macro_name, arity} = tuple, def_kind)
+      when is_atom(module) and is_atom(function_macro_name) and
+           is_integer(arity) and arity >= 0 and arity <= 255 and
+           def_kind in [:def, :defp, :defmacro, :defmacrop] do
     assert_not_compiled!(:defines?, module)
     table = defs_table_for(module)
     case :ets.lookup(table, {:def, tuple}) do
-      [{_, ^kind, _, _, _, _}] -> true
+      [{_, ^def_kind, _, _, _, _}] -> true
```
Comment (lexmag) [inline lib/elixir/lib/module.ex]: Perhaps we should keep `kind`, we didn't change that in `add_doc` as well.

File: lib/elixir/lib/module.ex [hunk 28]
```diff
@@ -788,10 +809,12 @@ defmodule Module do
...
 
   """
-  def definitions_in(module, kind) do
+  @spec definitions_in(module, def_kind) :: [function_arity]
+  def definitions_in(module, def_kind)
+      when is_atom(module) and def_kind in [:def, :defp, :defmacro, :defmacrop] do
     assert_not_compiled!(:definitions_in, module)
     table = defs_table_for(module)
-    :lists.concat :ets.match(table, {{:def, :'$1'}, kind, :_, :_, :_, :_})
+    :lists.concat :ets.match(table, {{:def, :'$1'}, def_kind, :_, :_, :_, :_})
```
Comment (lexmag) [inline lib/elixir/lib/module.ex]: Same as https://github.com/elixir-lang/elixir/pull/5869/files#r107312839.

File: lib/elixir/lib/module.ex (hunk 26 above)
Comment (lexmag) [inline lib/elixir/lib/module.ex]: Indeed, we use `function_name` in same context in `make_overridable`, and `overridable?`.

File: lib/elixir/lib/module.ex [hunk 29]
```diff
@@ -801,42 +824,50 @@ defmodule Module do
...
+              "cannot make function #{function_name}/#{arity} overridable because it was not defined"
+          clause ->
+            neighbours =
+              if :elixir_compiler.get_opt(:internal) do
+                []
+              else
+                Module.LocalsTracker.yank(module, tuple)
+              end
+
+            old  = :elixir_overridable.overridable(module)
```
Comment (lexmag) [inline lib/elixir/lib/module.ex]: Extra space here to ✂️.

File: lib/elixir/test/elixir/module_test.exs [hunk 30]
```diff
@@ -310,4 +316,19 @@ defmodule ModuleTest do
       assert Module.definitions_in(__MODULE__, :defp) == []
     end
   end
+
+  test "make_overridable/2 with bad arguments" do
```
Comment (lexmag) [inline lib/elixir/test/elixir/module_test.exs]: I think we usually use `invalid` instead of `bad`. :bowtie:

File: lib/elixir/test/elixir/module_test.exs [hunk 31]
```diff
@@ -50,11 +50,17 @@ defmodule ModuleTest do
   end
   Module.eval_quoted __MODULE__, contents, [], file: "sample.ex", line: 13
 
+  defp purge(modules) do
+    Enum.each modules, fn(m) ->
```
Comment (lexmag) [inline lib/elixir/test/elixir/module_test.exs]: Seems premature use of `Enum` here, in places where we call `purge/1` helper we pass only one module.

File: lib/elixir/lib/module.ex [hunk 32]
```diff
@@ -944,18 +975,20 @@ defmodule Module do
       end
 
   """
-  def register_attribute(module, new, opts) when is_atom(new) do
+  @spec register_attribute(module, attribute :: atom, opts :: [{:accumulate, boolean}, {:persist, boolean}])
+      :: :ok
+  def register_attribute(module, attribute, opts) when is_atom(module) and is_atom(attribute) do
```
Comment (lexmag) [inline lib/elixir/lib/module.ex]: I'm wondering if we're fine to rename `opts` –> `options`?

File: lib/elixir/lib/module.ex [hunk 33]
```diff
@@ -351,6 +351,10 @@ defmodule Module do
   the documentation for the [`:compile` module](http://www.erlang.org/doc/man/compile.html).
   '''
 
+  @typep function_arity :: {atom, arity}
```
Comment (lexmag) [inline lib/elixir/lib/module.ex]: I think we can name it `definition`, as it this name is already used in `definitions_in` function, which returns in exactly that format.

File: lib/elixir/lib/module.ex (hunk 26 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: I am worried that `function_name` doesn't take macros. That's why the difference.
We could name it `function_or_macro_name`

File: lib/elixir/lib/module.ex (hunk 27 above)
Comment (eksperimental) [inline lib/elixir/lib/module.ex]: this is done intentionally, to be inlined with the `typep`s:
in add_doc is called `kind` because its typespec is a union `def_kind | type_kind`

**Comment by eksperimental** [comment]
@lexmag I have added a commit with your suggestions. I didn't add implement since I have left a comment waiting for your answer.

File: lib/elixir/lib/module.ex (hunk 26 above)
Comment (whatyouhide) [inline lib/elixir/lib/module.ex]: 👍 for clarity, let's go with `function_or_macro_name` :)

Comment (eksperimental) [inline lib/elixir/lib/module.ex]: fixed

**Comment by eksperimental** [comment]
this is ready to merge.

**Comment by whatyouhide** [comment]
Thanks @eksperimental! 💟

**Comment by eksperimental** [comment]
thank you everyone (@elixir-lang/elixir)  for your suggestions!
